api/src/main/kotlin/org/jellyfin/apiclient/api/sockets/WebSocketApi.kt

```diff
@@ -0,0 +1,273 @@
+			"RefreshProgress" to serializer<RefreshProgressMessage>(),
+			"ScheduledTaskEnded" to serializer<ScheduledTaskEndedMessage>(),
+			"PackageInstallationCancelled" to serializer<PackageInstallationCancelledMessage>(),
+			"PackageInstallationFailed" to serializer<PackageInstallationFailedMessage>(),
+			"PackageInstallationCompleted" to serializer<PackageInstallationCompletedMessage>(),
+			"PackageInstalling" to serializer<PackageInstallingMessage>(),
+			"PackageUninstalled" to serializer<PackageUninstalledMessage>(),
+			"ActivityLogEntry" to serializer<ActivityLogEntryMessage>(),
+			"ScheduledTasksInfo" to serializer<ScheduledTasksInfoMessage>(),
+
+			// Shared type, only implemented as outgoing message
+			"KeepAlive" to serializer<KeepAliveMessage>(),
+		)
+
+		private const val JSON_PROP_DATA = "Data"
+		private const val JSON_PROP_MESSAGE_ID = "MessageId"
+		private const val JSON_PROP_MESSAGE_TYPE = "MessageType"
+		private const val RECONNECT_DELAY = 3 * 1000L // milliseconds
+	}
+
+	private val logger = LoggerFactory.getLogger("WebSocketApi")
+
+	private val json = Json {
+		// Based on DefaultJson
+		isLenient = false
+		ignoreUnknownKeys = true
+		allowSpecialFloatingPointValues = true
+		useArrayPolymorphism = false
+		encodeDefaults = true
+	}
+
+	private var socketJob: Job? = null
+	private val subscriptions = mutableListOf<SocketSubscription>()
+
+	private val outgoingMessageChannel by lazy {
+		Channel<String>(BUFFERED)
+	}
+
+	private suspend fun ReceiveChannel<Frame>.read() = consumeAsFlow()
+		.onEach {
+			val message = it.readSocketMessage() ?: return@onEach
+
+			// Send message to subscriptions
+			subscriptions.forEach { subscription -> subscription.callback(message) }
+		}
+		.catch { it.printStackTrace() }
+		.onCompletion {
+			// Reconnect
+			logger.debug("Socket receiver completed, found %s subscriptions", subscriptions.size)
+			delay(RECONNECT_DELAY)
+			if (subscriptions.isNotEmpty()) reconnect()
+		}
+		.collect()
+
+	private suspend fun SendChannel<Frame>.write() = outgoingMessageChannel
+		.receiveAsFlow()
+		.onEach { text ->
+			logger.info("Sending message {}", text)
+			send(Frame.Text(text))
+		}
+		.catch { it.printStackTrace() }
+		.collect()
+
+	private suspend fun subscriptionsChanged() {
+		logger.debug("Subscriptions changed to %s", subscriptions.size)
+
+		if (socketJob != null && subscriptions.isEmpty()) {
+			logger.info("Dropping connection")
+			socketJob?.cancel()
+		} else if (socketJob == null && subscriptions.isNotEmpty()) {
+			logger.info("Creating connection")
+			reconnect()
+		}
+	}
+
+	/**
+	 * Call to (re)connect the WebSocket. Does not close current listeners.
+	 */
+	suspend fun reconnect() {
+		// Get access token and check if it's not null
+		val accessToken = checkNotNull(api.accessToken)
+
+		logger.debug("Reconnect requested")
+
+		// Close existing socket
+		socketJob?.cancel()
+
+		// Open socket
+		socketJob = GlobalScope.launch {
```

Comment:
Normally a coroutine will be bound to it's parent coroutine so they will auto-cancel and stuff. Unfortunately this won't work with how I want to make this class work, because it would bind to the scope calling the `subscribe` function, which can end early.

AFAIK this is a valid use-case to use the GlobalScope which binds the coroutine to the application, which is dangerous if you don't keep track of it. I track the subscriptions and if they reach 0 I will cancel this job though (See `subscriptionsChanged`).

I'm still not an expert in Kotlin's coroutines, they are quite complex. But in my testing this worked as expected.